README.md
# Marketplace Partner Tools

[![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)](LICENSE)
[![Pull Requests Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat)](http://makeapullrequest.com)

Image validation, automation, and other tools for DigitalOcean Marketplace partners and Custom Image users.

![Screenshot](screenshot.png)

This repository includes bash scripts intended for use by Marketplace partners and Custom Image users to validate their images prior to submission, or use.

## Build an Image Manually

If you're just getting started, [build-an-image.md](marketplace_docs/build-an-image.md) is where to begin. Some things here are required in your Fabric automation.

## Build an Image with Fabric Automation

For repeatable image build processing, [build-an-image-fabric.md](marketplace_docs/build-an-image-fabric.md) is a good next step.

## marketplace_validation

Provides a script `img_check.sh` that validates your build Droplet before shutting down to [create a snapshot](https://www.digitalocean.com/docs/images/snapshots/how-to/snapshot-droplets/).

### Supported Distributions 

* Debian 9.x (stretch)
* Ubuntu 18.04 (LTS)
* Ubuntu 16.04 (LTS)
* CentOS 7.x
* CentOS 6.x 

This script checks build Droplet images against the following criteria:

- Check for a supported distro and release
- Check that all security updates are installed
- Check for a populated bash_history file
- Check for ssh_keys in the image
- Check for pre-set passwords
- Check that packages are updated
- Check that a firewall is installed/configured
- Check that a valid version of cloud-init is installed
- Check for any populated log files or log archives

*coming soon* - Allow the script to be run against a disk image file on a Linux workstation by mounting and chrooting into the image.

The script img_check.sh can be used from within the image build system prior to [creating a snapshot](https://www.digitalocean.com/docs/images/snapshots/how-to/snapshot-droplets/) or exporting a disk image from Virtualbox, VMWare or a physical machine. The script is minimally invasive and designed to avoid changes to the disk. The one exception is that this script will update your apt or yum package database in order to identify if there are any uninstalled security updates.

To clear log files, bash_history, and authorized keys, you can use any file clearing method, including truncate against each result the script outputs, e.g.,

`$ truncate --size 0 <path/to/file.log>`


CHANGELOG.md
## CHANGELOG



### Add Support for Debian 9

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 29 Mar 2019 18:06:36 +0000  
> Author: root (root@debian-s-1vcpu-1gb-nyc1-01.localdomain)  
> Committer: root (root@debian-s-1vcpu-1gb-nyc1-01.localdomain)  

- add support to image test script for Debian
- update README's
- for more info on changelog hooks please ref: https://github.com/dOpensource/dsiprouter/tree/dev/resources/git


---


### Update build-an-image-fabric.md

> Branches Affected: master  
> Tags Affected:   
> Date: Mon, 25 Mar 2019 15:41:52 +0100  
> Author: Divyendu Singh (divyendu.z@gmail.com)  
> Committer: GitHub (noreply@github.com)  



---


### IMG-535 - Ensure validation script exits with proper status codes

> Branches Affected: master  
> Tags Affected:   
> Date: Thu, 21 Mar 2019 13:19:37 -0500  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### Add step to cleanup to clear root mailbox if any mail is present

> Branches Affected: master  
> Tags Affected:   
> Date: Thu, 21 Mar 2019 13:13:49 -0500  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### also allow CentOs convention for locked password on root

> Branches Affected: master  
> Tags Affected:   
> Date: Mon, 18 Mar 2019 18:40:25 -0400  
> Author: Ken Bingham (w@qrk.us)  
> Committer: Ken Bingham (w@qrk.us)  



---


### img_check.sh: Print date that script is run

> Branches Affected: master  
> Tags Affected:   
> Date: Wed, 13 Mar 2019 16:03:30 -0400  
> Author: John Gannon (jgannon@digitalocean.com)  
> Committer: GitHub (noreply@github.com)  

Assists in debugging efforts that may come up between DO and Vendors, as well as in DO's listing submission process.


---


### build-an-image.md: doctl instructions added

> Branches Affected: master  
> Tags Affected:   
> Date: Wed, 13 Mar 2019 14:40:54 -0400  
> Author: John Gannon (jgannon@digitalocean.com)  
> Committer: GitHub (noreply@github.com)  



---


### removed .backup file

> Branches Affected: master  
> Tags Affected:   
> Date: Mon, 11 Mar 2019 11:17:59 -0500  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### Added Packer example and reorganized template folders

> Branches Affected: master  
> Tags Affected:   
> Date: Mon, 11 Mar 2019 11:16:37 -0500  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### update img_check.sh exclude lfd.log log

> Branches Affected: master  
> Tags Affected:   
> Date: Wed, 6 Mar 2019 23:25:48 +1000  
> Author: George Liu (eva2000@centminmod.com)  
> Committer: George Liu (eva2000@centminmod.com)  



---


### img_check.sh centos 7 & csf firewall compatibility

> Branches Affected: master  
> Tags Affected:   
> Date: Wed, 6 Mar 2019 22:00:27 +1000  
> Author: George Liu (eva2000@centminmod.com)  
> Committer: George Liu (eva2000@centminmod.com)  



---


### README.md: Link out to snapshots documentation

> Branches Affected: master  
> Tags Affected:   
> Date: Mon, 4 Mar 2019 16:38:06 -0500  
> Author: John Gannon (jgannon@digitalocean.com)  
> Committer: GitHub (noreply@github.com)  



---


### Update build-an-image-fabric.md

> Branches Affected: master  
> Tags Affected:   
> Date: Thu, 28 Feb 2019 12:03:14 -0500  
> Author: Rijk van Zanten (rijkvanzanten@me.com)  
> Committer: GitHub (noreply@github.com)  

Hey there! I ran into some versioning issues while working with the example project. I hope this note will make sure other people don't spend 30min messing around as I did 🙂


---


### Update 001_onboot

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 22 Feb 2019 14:29:46 -0600  
> Author: Ryan Quinn (ryan@digitalocean.com)  
> Committer: GitHub (noreply@github.com)  

Fix for typo reported by @kspearrin


---


### Fix for IMG-559 / Issue 16 - Fail if private key found

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 22 Feb 2019 08:16:47 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### remove shutdown func add exit to build func and add copy for user

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 15 Feb 2019 10:30:09 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### stop apt-get from prompting about config files when upgrading packages

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 15 Feb 2019 10:04:01 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### replace per-once with per-instance

> Branches Affected: master  
> Tags Affected:   
> Date: Fri, 15 Feb 2019 09:57:35 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### Fix to change recommendation from per-once to per-instance on firstboot scripts

> Branches Affected: master  
> Tags Affected:   
> Date: Wed, 13 Feb 2019 13:14:51 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### Ensure instance folder is removed on build system

> Branches Affected: master  
> Tags Affected:   
> Date: Tue, 12 Feb 2019 15:05:55 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### Fix for permissions not being copied and adding noninteractive env var

> Branches Affected: master  
> Tags Affected:   
> Date: Tue, 12 Feb 2019 15:04:53 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### moved postfix install fix to run if uncommented before apt installs

> Branches Affected: master  
> Tags Affected:   
> Date: Tue, 12 Feb 2019 12:41:02 -0600  
> Author: rquinn (ryan@digitalocean.com)  
> Committer: rquinn (ryan@digitalocean.com)  



---


### build-an-image.md: typo


byoi_validation/img_check.sh
#!/bin/bash
#
# DigitalOcean BYOI Image Validation Tool
# © 2018 DigitalOcean LLC.
# This code is licensed under MIT license (see LICENSE.txt for details)
#
VERSION="v. 0.1"

# Script should be run with SUDO
if [ "$EUID" -ne 0 ]
  then echo "[Error] - This script must be run with sudo or as the root user."
  exit
fi

STATUS=0
PASS=0
WARN=0
FAIL=0

function getDistro {
    if [ -f /etc/os-release ]; then
    # freedesktop.org and systemd
    . /etc/os-release
    OS=$NAME
    VER=$VERSION_ID
elif type lsb_release >/dev/null 2>&1; then
    # linuxbase.org
    OS=$(lsb_release -si)
    VER=$(lsb_release -sr)
elif [ -f /etc/lsb-release ]; then
    # For some versions of Debian/Ubuntu without lsb_release command
    . /etc/lsb-release
    OS=$DISTRIB_ID
    VER=$DISTRIB_RELEASE
elif [ -f /etc/debian_version ]; then
    # Older Debian/Ubuntu/etc.
    OS=Debian
    VER=$(cat /etc/debian_version)
elif [ -f /etc/SuSe-release ]; then
    # Older SuSE/etc.
    :
elif [ -f /etc/redhat-release ]; then
    # Older Red Hat, CentOS, etc.
    VER=$( cat /etc/redhat-release | cut -d" " -f3 | cut -d "." -f1)
    d=$( cat /etc/redhat-release | cut -d" " -f1 | cut -d "." -f1)
    if [[ $d == "CentOS" ]]; then
      OS="CentOS Linux"
    fi
else
    # Fall back to uname, e.g. "Linux <version>", also works for BSD, etc.
    OS=$(uname -s)
    VER=$(uname -r)
fi
}
function loadPasswords {
SHADOW=$(cat /etc/shadow)
}
function checkLogs {
    echo "Checking for log files which may contain sensitive information or details you may not wish to share.  Please review any issues detected to ensure that the contents of these files are not sensitive."
    echo -en "\nChecking for log files in /var/log\n\n"
    # Check if there are log archives or log files that have not been recently cleared.
    for f in /var/log/*-????????; do
      [[ -e $f ]] || break
      echo -en "\e[93m[WARN]\e[0m Log archive ${f} found\n"
      ((WARN++))
      if [[ $STATUS != 2 ]]; then
        
          STATUS=1
      fi
    done
    for f in  /var/log/*.[0-9];do
      [[ -e $f ]] || break
      echo -en "\e[93m[WARN]\e[0m Log archive ${f} found\n"
      ((WARN++))
      if [[ $STATUS != 2 ]]; then
        
          STATUS=1
      fi
    done
    for f in /var/log/*.log; do
      [[ -e $f ]] || break
      if [ "$( cat "${f}" | wc -c)" -gt 50 ]; then
          echo -en "\e[93m[WARN]\e[0m un-cleared log file, ${f} found\n"
          ((WARN++))
          if [[ $STATUS != 2 ]]; then
        
          STATUS=1
      fi
      fi
    done
}
function checkTMP {
  # Check the /tmp directory to ensure it is empty.  Warn on any files found.
  return 1
}
function checkRoot {
    user="root"
    uhome="/root"
    for usr in $SHADOW
    do
      IFS=':' read -r -a u <<< "$usr"
      if [[ "${u[0]}" == "${user}" ]]; then
        if [[ ${u[1]} == "!" ]] || [[ ${u[1]} == "*" ]]; then
            echo -en "\e[32m[PASS]\e[0m User ${user} has no password set.\n"
            ((PASS++))
        else
            echo -en "\e[93m[WARN]\e[0m User ${user} has a password set on their account.  This password will be included on Droplets created from this image\n"
            ((WARN++))
            if [[ $STATUS != 2 ]]; then          
                STATUS=1
            fi
        fi
      fi
    done
    if [ -d ${uhome}/ ]; then
            if [ -d ${uhome}/.ssh/ ]; then
                if  ls ${uhome}/.ssh/*> /dev/null 2>&1; then
                  echo "Checking for ssh-keys in user directories... authorized_keys files list keys that are allowed access to Droplets created from this image."
                  
                    for key in ${uhome}/.ssh/*
                        do
                             if  [ "${key}" == "${uhome}/.ssh/authorized_keys" ]; then
                            
                                if [ "$( cat "${key}" | wc -c)" -gt 50 ]; then
                                    echo -en "\e[41m[FAIL]\e[0m User \e[1m${user}\e[0m has a populated authorized_keys file in \e[93m${key}\e[0m\n"
                                    ((FAIL++))
                                    STATUS=2
                                fi
                            
                            elif  [ "${key}" != "${uhome}/.ssh/known_hosts" ]; then
                                
                                 echo -en "\e[93m[WARN]\e[0m User \e[1m${user}\e[0m has a file in their .ssh directory at \e[93m${key}\e[0m\n"
                                    ((WARN++))
                                    if [[ $STATUS != 2 ]]; then
                                      
                                        STATUS=1
                                    fi
                            else
                                if [ "$( cat "${key}" | wc -c)" -gt 50 ]; then
                                    echo -en "\e[93m[WARN]\e[0m User \e[1m${user}\e[0m has a populated known_hosts file in \e[93m${key}\e[0m\n"
                                    ((WARN++))
                                    if [[ $STATUS != 2 ]]; then
                                      
                                        STATUS=1
                                    fi
                                fi
                            fi
                        done
                else
                    echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m has no SSH keys present\n"
                    
                fi
            else
                echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m does not have an .ssh directory\n"
            fi
             if [ -f /root/.bash_history ];then
        
                      BH_S=$( cat /root/.bash_history | wc -c)
                  
                      if [[ $BH_S -lt 200 ]]; then
                          echo -en "\e[32m[PASS]\e[0m ${user}'s Bash History appears to have been cleared\n"
                          ((PASS++))
                      else
                          echo -en "\e[93m[WARN]\e[0m ${user}'s Bash History should be cleared to prevent sensitive information from leaking\n"
                          ((WARN++))
                            
                              if [[ $STATUS != 2 ]]; then
                    
                                  STATUS=1
                              fi
                          
                      fi
                      
                      return 1;
                  else
                      echo -en "\e[32m[PASS]\e[0m The Root User's Bash History is not present\n"
                      ((PASS++))
                  fi
        else
            echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m does not have a directory in /home\n"
            
        fi
        echo -en "\n\n"
    return 1
}

function checkUsers {
    # Check each user-created account
    for user in $(awk -F: '$3 >= 1000 && $1 != "nobody" {print $1}' /etc/passwd;)
    do
      
      # Skip some other non-user system accounts
      if [[ $user == "centos" ]]; then
        :
      elif [[ $user == "nfsnobody" ]]; then
        :
    else
      echo -en "\nChecking user: ${user}...\n"
      for usr in $SHADOW
        do
          IFS=':' read -r -a u <<< "$usr"
          if [[ "${u[0]}" == "${user}" ]]; then
              if [[ ${u[1]} == "!" ]] || [[ ${u[1]} == "*" ]]; then
                  echo -en "\e[32m[PASS]\e[0m User ${user} has no password set.\n"
                  ((PASS++))
              else
                  echo -en "\e[93m[WARN]\e[0m User ${user} has a password set on their account.\n"
                  ((WARN++))
                  if [[ $STATUS != 2 ]]; then
                    
                      STATUS=1
                  fi
              fi
          fi
        done
      
      
      
       
        #echo "User Found: ${user}"
        uhome="/home/${user}"
        if [ -d "${uhome}/" ]; then
            if [ -d "${uhome}/.ssh/" ]; then
                if  ls "${uhome}/.ssh/*"> /dev/null 2>&1; then
                  
                    for key in ${uhome}/.ssh/*
                        do
                         
                            if  [ "${key}" == "${uhome}/.ssh/authorized_keys" ]; then
                            
                                if [ "$( cat "${key}" | wc -c)" -gt 50 ]; then
                                    echo -en "\e[41m[FAIL]\e[0m User \e[1m${user}\e[0m has a populated authorized_keys file in \e[93m${key}\e[0m\n"
                                    ((FAIL++))
                                    STATUS=2
                                fi
                            elif  [ "${key}" != "${uhome}/.ssh/known_hosts" ]; then
                                
                                 echo -en "\e[93m[WARN]\e[0m User \e[1m${user}\e[0m has a file in their .ssh directory named \e[93m${key}\e[0m\n"
                                 ((WARN++))
                                 if [[ $STATUS != 2 ]]; then
                                        STATUS=1
                                    fi
                            
                            else
                                if [ "$( cat "${key}" | wc -c)" -gt 50 ]; then
                                    echo -en "\e[93m[WARN]\e[0m User \e[1m${user}\e[0m has a known_hosts file in \e[93m${key}\e[0m\n"
                                    ((WARN++))
                                    if [[ $STATUS != 2 ]]; then
                                        STATUS=1
                                    fi
                                fi
                            fi
                            
                           
                        done
                else
                    echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m has no SSH keys present\n"
                fi
            else
                echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m does not have an .ssh directory\n"
            fi
        else
            echo -en "\e[32m[ OK ]\e[0m User \e[1m${user}\e[0m does not have a directory in /home\n"
        fi
        
         # Check for an uncleared .bash_history for this user
              if [ -f "${uhome}/.bash_history" ]; then
                            BH_S=$( cat "${uhome}/.bash_history" | wc -c )
    
                            if [[ $BH_S -lt 200 ]]; then
                                echo -en "\e[32m[PASS]\e[0m ${user}'s Bash History appears to have been cleared\n"
                                ((PASS++))
                            else
                                echo -en "\e[93m[WARN]\e[0m ${user}'s Bash History should be cleared to prevent sensitive information from leaking\n"
                                ((WARN++))
                                    if [[ $STATUS != 2 ]]; then
                                        STATUS=1
                                    fi
                                
                            fi
                           echo -en "\n\n"
                         fi
        fi
    done
}
function checkFirewall {
    
    if [[ $OS == "Ubuntu" ]]; then
      fw="ufw"
      service ufw status >/dev/null 2>&1
    elif [[ $OS == "CentOS Linux" ]]; then
      fw="firewalld"
      systemctl status firewalld >/dev/null 2>&1
    fi
    
    if [ $? = 0 ]; then
        FW_VER="\e[32m[PASS]\e[0m Firewall service (${fw}) is active\n"
        ((PASS++))
    else
         FW_VER="\e[93m[WARN]\e[0m No firewall is configured. Ensure ${fw} is installed and configured\n"
          ((WARN++))
        if [[ $STATUS != 2 ]]; then
            STATUS=1
        fi
    fi
    
}
function checkUpdates {
    if [[ $OS == "Ubuntu" ]]; then
        echo -en "\nUpdating apt package database to check for security updates, this may take a minute...\n\n"
        apt-get -y update > /dev/null
        update_count=$(/usr/lib/update-notifier/apt-check 2>&1 | cut -d ';' -f 2)
        if [[ $update_count -gt 0 ]]; then
            echo -en "\e[41m[FAIL]\e[0m There are ${update_count} security updates available for this image that have not been installed.\n"
            ((FAIL++))
            STATUS=2
        else
            echo -en "\e[32m[PASS]\e[0m There are no pending security updates for this image.\n\n"
        fi
    elif [[ $OS == "CentOS Linux" ]]; then
        echo -en "\nChecking for available updates with yum, this may take a minute...\n\n"
        
        update_count=$(yum list updates -q | grep -vc "Updated Packages")
         if [[ $update_count -gt 0 ]]; then
            echo -en "\e[41m[FAIL]\e[0m There are ${update_count} updates available for this image that have not been installed.\n"
            ((FAIL++))
            STATUS=2
        else
            echo -en "\e[32m[PASS]\e[0m There are no pending security updates for this image.\n"
            ((PASS++))
        fi
    else
        echo "Error encountered"
        exit
    fi

    return 1;    
}
function checkCloudInit {
    
    if hash cloud-init 2>/dev/null; then
        CI="\e[32m[PASS]\e[0m Cloud-init is installed.\n"
        CI_VER=$(cloud-init -v)
        ((PASS++))
    else
        CI="\e[93m[WARN]\e[0m No valid verison of cloud-init was found. Images without cloud-init are not officially supported.\n"
        ((WARN++))
        if [[ $STATUS != 2 ]]; then
                                      
            STATUS=1
        fi
    fi    
    return 1
}

clear
echo "DigitalOcean BYOI Image Validation Tool ${VERSION}"
echo "This script currently supports the following Linux distributions:"
echo " - Ubuntu 16.04 LTS"
echo " - Ubuntu 18.04 LTS"
echo " - CentOS 7.x"
echo "Checking local system for BYOI compatibility..."

getDistro

echo -en "\n\e[1mDistribution:\e[0m ${OS}\n"
echo -en "\e[1mVersion:\e[0m ${VER}\n\n"

ost=0
osv=0

if [[ $OS == "Ubuntu" ]]; then
        ost=1
    if [[ $VER == "18.04" ]]; then
        osv=1
    elif [[ $VER == "16.04" ]]; then
        osv=1
    else
        osv=0
    fi
    
elif [[ $OS == "CentOS Linux" ]]; then
        ost=1
     if [[ $VER == "7" ]]; then
        osv=1
    else
        osv=2
    fi
else
    ost=0
fi

if [[ $ost == 1 ]]; then
    echo -en "\e[32m[PASS]\e[0m Supported Operating System Detected: ${OS}\n"
    ((PASS++))
else
    echo -en "\e[41m[FAIL]\e[0m ${OS} is not a supported Operating System.  This utility is not intended for use with this operating system\n"
    ((FAIL++))
    STATUS=2
fi

if [[ $osv == 1 ]]; then
    echo -en "\e[32m[PASS]\e[0m Supported Release Detected: ${VER}\n"
    ((PASS++))
elif [[ $ost == 1 ]]; then
    echo -en "\e[41m[FAIL]\e[0m ${OS} ${VER} is not a supported Operating System Version. This utility has not been tested with this version of ${OS}.\n"
    ((FAIL++))
    STATUS=2
else
    echo "Exiting..."
    exit
fi

checkCloudInit

echo -en "\n\n"
echo "Cloud-init is required for several features of the DigitalOcean cloud.  We strongly recommend using"
echo "a modern version of cloud init (0.77 or greater).  Versions as early as 0.75 will provide partial"
echo "support but some DigitalOcean products or features may be unavailable."

echo -en "\n${CI}\n"
if [[ $CI_VER ]]; then
echo -en "\n\nThis system has cloud init version: "
echo -en "\n\n\\e[36m${CI_VER}\e[0m\n\n"
else
echo en "\n\nCloud init was not detected.  Please install it with 'apt install cloud-init' (Ubuntu) or 'yum install cloud-init' (CentOS)"
fi
checkFirewall

echo -en "${FW_VER}"

checkUpdates

loadPasswords

checkLogs

echo -en "\n\nChecking all user-created accounts...\n"
checkUsers

echo -en "\n\nChecking the root account...\n"
checkRoot


# Summary
echo -en "\n\n---------------------------------------------------------------------------------------------------\n"

if [[ $STATUS == 0 ]]; then
    echo -en "Scan Complete.\n\e[32mAll Tests Passed!\e[0m\n"
elif [[ $STATUS == 1 ]]; then
    echo -en "Scan Complete. \n\e[93mSome non-critical tests failed.  Please review these items.\e[0m\e[0m\n"
else
    echo -en "Scan Complete. \n\e[41mOne or more tests failed.  Please review these items and re-test.\e[0m\n"
fi
echo "---------------------------------------------------------------------------------------------------"
echo -en "\e[1m${PASS} Tests PASSED\e[0m\n"
echo -en "\e[1m${WARN} WARNINGS\e[0m\n"
echo -en "\e[1m${FAIL} Tests FAILED\e[0m\n"

echo -en "---------------------------------------------------------------------------------------------------\n"

if [[ $STATUS == 0 ]]; then
    echo -en "We did not detect any issues with this image. Please be sure to manually ensure that all software installed on the base system is functional, secure and properly configured(or facilities for configuration on first-boot have been created).\n\n"
elif [[ $STATUS == 1 ]]; then
    echo -en "Please review all [WARN] items above and ensure they are intended or resolved.  If you do not have a specific requirement, we recommend resolving these items before using this image in production\n\n"
else
    echo -en "Some critical tests failed.  This means your image should be reviewed manually or this test should be re-run before use in production\n\n"
fi
